### PR TITLE
Rust matmul from Hackweek

### PR DESCRIPTION
**Pull request by mhauru**
I simplified the tensor code we wrote in Rust during the hackweek, removing everything about multidimensional arrays and everything that wasn't serving the purpose of multiplying matrices, and turned it into a benchmark competitor.

### CODE REVIEW

**Comment by edchapman88** [comment]
Nice one! Presumably we wait with bated breath until someone finds the time to run the benchmark?

**Comment by mhauru** [comment]
It finishes in about 2.3s on my Macbook, but I think @llewelld has some standardised set up to run them.

**Comment by llewelld** [comment]
Great! I can run the Intel i7 and M1 Pro benchmarks with the same configuration as the other implementations. I'll post a review for this PR once I've run the tests, but will probably create a separate PR with the results.

The M2 benchmarks for the other implementations were run by @triangle-man so ideally he should also run these for a full set.

File: implementation/matmul-rst-hackweek/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+impl<E: Element> MatrixImpl<E> {
+    /// Multiply self by the transpose of a matrix.
+    ///
+    /// This is equivalent to `self.matmul(other.transpose())`, but faster.
+    fn matmul_transpose(&self, other: &Self) -> MatrixImpl<E> {
+        debug_assert!(self.num_cols != other.num_cols);
+        let dim1 = self.num_rows;
+        let dim_inner = self.num_cols;
+        let dim2 = other.num_rows;
+        // Create an unallocated data vector for the result.
```
Comment (llewelld) [inline implementation/matmul-rst-hackweek/src/lib.rs:99]:
```suggestion
        // Create an uninitialised data vector for the result.
```
```suggestion
        // Create an pre-allocated data vector for the result.
```
This looks like a typo to me: the vector is intentionally being allocated in advance.

File: implementation/matmul-rst-hackweek/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+        };
+    }
+
+    /// Multiply self by a matrix.
+    fn matmul_straight(&self, other: &Self) -> MatrixImpl<E> {
+        debug_assert!(self.num_cols != other.num_rows);
+        let dim1 = self.num_rows;
+        let dim_inner = self.num_cols;
+        let dim2 = other.num_cols;
+        // Create an unallocated data vector for the result.
```
Comment (llewelld) [inline implementation/matmul-rst-hackweek/src/lib.rs:134]:
```suggestion
        // Create an uninitialised data vector for the result.
```
```suggestion
        // Create an pre-allocated data vector for the result.
```
Similarly.

**Review by llewelld — APPROVED** [review]
There are comments for just a couple of super-minor points. Otherwise it looks great.

I've added the benchmark results in #46, pointing to this branch in case you want to merge that in here before merging everything into main. Otherwise we can merge that into main afterwards.

It'd be neat to also add this implementation to the [results for scaling up larger matrices](https://github.com/alan-turing-institute/hut23-linear-algebra/tree/main/implementation#benchmarking-matrix-size-scaling-results) so that the transpose approach gets run as well (not in this PR though).

**Pull request merged**